mixin-metrics.go

```diff
@@ -0,0 +1,319 @@
+	for k := range metrics {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	errStrings := make([]string, 0, len(errs))
+	for _, err := range errs {
+		errStrings  = append(errStrings, err.Error())
+	}
+
+	metricsFile := MetricsFile{
+		Filename: fn,
+		Metrics: keys,
+		ParseErrors: errStrings,
+	}
+
+	return metricsFile
+}
+
+// use promql parser on a query to extract metrics
+func ParseQuery(query string, metrics map[string]struct{}) error {
+
+	query = strings.ReplaceAll(query, `\"`, `"`)
+	query = strings.ReplaceAll(query, `\n`, ``)
+	query = strings.ReplaceAll(query, `$__interval`, "5m")
+	query = strings.ReplaceAll(query, `$__rate_interval`, "5m")
+	query = strings.ReplaceAll(query, `$interval`, "5m")
+	query = strings.ReplaceAll(query, `$resolution`, "5s")
+
+	// label_values 
+	if strings.Contains(query, "label_values"){
+		re := regexp.MustCompile(`label_values\(([a-zA-Z0-9_]+)`)
+		query = re.FindStringSubmatch(query)[1]
+	}
+
+	expr, err := parser.ParseExpr(query)
+	if err != nil {
+		err = errors.Wrapf(err, "promql query=%v", query)
+		log.Debugln("msg", "promql parse error", "err", err, "query", query)
+		return err
+	}
+
+	parser.Inspect(expr, func(node parser.Node, path []parser.Node) error {
+		if n, ok := node.(*parser.VectorSelector); ok {
+			metrics[n.Name] = struct{}{}
+		}
+		return nil
+	})
+
+	return nil
+}
+
+// use jq on a file to extract prom queries
+func ParseJq(queries *[]string, jsonData map[string]interface{}, jqExpr string) error {
+
+	query, err := gojq.Parse(jqExpr)
+	if err != nil {
+		return err
+	}
+
+	iter := query.Run(jsonData)
+	for {
+		v, ok := iter.Next()
+		if !ok {
+			break
+		}
+		if err, ok := v.(error); ok {
+			return err
+		}
+		*queries = append(*queries, v.(string))
+	}
+
+	return nil
+}
+
+// parses through a dashboard to extract queries, and then metrics from queries
+func ParseDashboard(file *os.File) (*MetricsFile, error) {
+
+	queries := make([]string, 0)
+	metrics := map[string]struct{}{}
+	errors := make([]error, 0)
+	res := map[string]interface{}{}
+
+	bytes, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, err
+	}
+
+	json.Unmarshal([]byte(bytes), &res)
+	if err := ParseJq(&queries, res, ".templating.list[].query"); err != nil {
+		return nil, err
+	}
+
+	if err := ParseJq(&queries, res, ".panels[]?.targets[]?.expr"); err != nil {
+		return nil, err
+	}
+
+	if err := ParseJq(&queries, res, ".rows[].panels[].targets[].expr"); err != nil {
+		return nil, err
+	}
+
+	for _, query := range queries {
+		if err := ParseQuery(query, metrics); err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	metricsFile := NewMetricsFile(file.Name(), metrics, errors)
+	return &metricsFile, nil
+}
+
+// todo: separate rules and raw metrics
+// parses through a rules file and extracts queries, and then metrics from queries
+func ParseRules(file *os.File) (*MetricsFile, error) {
+
+	metrics := map[string]struct{}{}
+	errors := make([]error, 0)
+
+	var conf RuleConfig
```

Comment:
nit: group variables and don't use `make` for an empty slice (the default value is enough)

```suggestion
    var (
        metrics = map[string]struct{}{}
      
        errors []error 
        conf RuleConfig 
    )
```